### PR TITLE
[AIMIGRAPHX-886] [AIMIGRAPHX-834] custom symbolic expression lib

### PR DESCRIPTION
**Pull request by shivadbhavsar**
## Motivation
A minimal symbolic expression library for MIGraphX. It provides construction, canonicalization, arithmetic, substitution/evaluation, parsing, and printing of symbolic integer expressions used in shape propagation

## Technical Details
Refer to the design document here: https://github.com/ROCm/AMDMIGraphX-internal/discussions/238

## Changelog Category

Add a `CHANGELOG.md` entry for any option other than `Not Applicable`
- - [x] Added: New functionality.
- - [ ] Changed: Changes to existing functionality.
- - [ ] Removed: Functionality or support that has been removed. (Compared to a previous release)
- - [ ] Optimized: Component performance that has been optimized or improved.
- - [ ] Resolved Issues: Known issues from a previous version that have been resolved.
- - [ ] Not Applicable: This PR is not to be included in the changelog.

### CODE REVIEW

**Review by CharlieL7 — CHANGES_REQUESTED** [review]
I asked Claude to examine the code for edge cases:
```
  Edge Cases Found



  1. Sum-times-sum is not expanded (biggest gap vs SymPy)

  make_mul only distributes integer * sum. Two sums become an opaque product:

   src/sym.cpp lines 393-404

  static expr_ptr make_mul(const expr_ptr& a, const expr_ptr& b)
  {
      if(holds<integer_data>(a) and holds<add_data>(b))
      {
          // ... distributes ...
      }
      if(holds<integer_data>(b) and holds<add_data>(a))
          return make_mul(b, a);
      auto pa = extract_mul(a);
      auto pb = extract_mul(b);
      // ... merges factor maps ...

  So (h+1)*(w+1) becomes mul_data{1, {(h+1)->1, (w+1)->1}}, not h*w + h + w + 1. This means mathematically equivalent expressions built different ways won't compare equal:

  auto a = (h+1) * (w+1);           // opaque product
  auto b = h*w + h + w + 1;         // expanded sum
  // a != b  (but mathematically they're equal)

  No test covers this. SymPy's expand() would fully expand it.


  2. Division by symbolic denominator -- no simplification

  make_floor_div only simplifies when the denominator is integer_data:

   src/sym.cpp lines 425-433

  static expr_ptr make_floor_div(const expr_ptr& a, const expr_ptr& b)
  {
      if(holds<integer_data>(b))
      {
          // ... simplification cases ...
      }
      return make_node(fdiv_data{a, b});
  }

  Consequences:
  • h / h does not simplify to 1
  • 2*h / h does not simplify to 2
  • 0 / h does not simplify to 0

  SymPy would simplify all three. No tests cover any of these.


  3. Division of sum by integer doesn't distribute

  (2*h + 4) / 2 is not simplified to h + 2. The code only checks if the numerator is mul_data (for coefficient cancellation) or integer_data, but not add_data where every term's coefficient is divisible:

   src/sym.cpp lines 436-441

          if(holds<mul_data>(a))
          {
              const auto& d = get_mul(a);
              if(d.coefficient % den == 0)
                  return build_mul(d.coefficient / den, d.factors);
          }

  SymPy would simplify (2*h + 4) / 2 to h + 2. Untested.


  4. Floor division semantics: C++ truncation vs mathematical floor

  The name says "floor division" but the implementation uses C++ truncation-toward-zero. For negative values these differ:

  ┌────────────┬─────────┬───────────────────────────┐
  │ Expression │ C++ (/) │ Python (//) / SymPy floor │
  ├────────────┼─────────┼───────────────────────────┤
  │ -7 / 2     │ -3      │ -4                        │
  │ 7 / -2     │ -3      │ -4                        │
  └────────────┴─────────┴───────────────────────────┘

  The test fdiv_negative_constant_folding tests against C++ semantics, so this is consistent internally. But if SymPy-style floor semantics are intended (and the name suggests they are), this is a bug. The eval_dim assert
  for non-negative results suggests dimensions should always be positive, which would mask this -- but subs() can produce intermediate negative values.


  5. `eval_dim` negative result only caught by assert


   src/sym.cpp lines 818-820

      auto v = eval_direct(p->node, bindings);
      assert(v >= 0 and "symbolic dimension evaluated to negative value");
      return static_cast<std::size_t>(v);

  In release builds (NDEBUG), a negative result silently wraps to a huge size_t. No test exercises this path, and there's no runtime protection.


  6. Integer overflow -- no protection

  All arithmetic is int64_t with no overflow checks. lit(INT64_MAX) + lit(1) or lit(INT64_MAX) * lit(2) is undefined behavior. SymPy uses arbitrary precision. No tests near int64 boundaries.


  7. Missing hash consistency tests

  Tests verify == works, but never verify the contract that a == b implies a.hash() == b.hash(). This is important since expr is used as an unordered_map key. A test like this is missing:

  auto a = h + w;
  auto b = w + h;
  EXPECT(a == b);
  EXPECT(a.hash() == b.hash());  // never tested



  8. Parser error paths untested

  The parser has multiple MIGRAPHX_THROW paths, none tested:
  • parse("") returns empty (tested implicitly via public API)
  • parse(")") -- unexpected character
  • parse("h +") -- unexpected end of input
  • parse("1 2") -- trailing characters
  • parse("-") alone -- unexpected end
  • parse("h**-1") -- negative exponent rejection
  • parse("h**h") -- non-integer exponent rejection
  • parse("--h") -- double unary minus (should work, producing h, but untested)



  9. Serialization gaps

  • No test for negative integers: round_trip(lit(-5))
  • No test for power expressions: round_trip(h * h) where the factor has exp=2
  • No test for negative coefficients: round_trip(-3 * h)
  • No test for fdiv with symbolic denominator: round_trip(h / w)



  10. `var("")` -- empty variable name

  make_symbol("") succeeds and creates a symbol with an empty name. This could collide with the empty string from expr{}.to_string(). Untested.


  11. `subs` that creates division by zero


  auto e = h / d;
  e.subs({{d, lit(0)}});  // throws during substitution

  The substitute function calls make_floor_div which throws on zero denominator. This is reasonable behavior but untested.


  12. `eval_dim` / `subs` with non-symbol key

  The public API validates that map keys are symbols and throws otherwise, but no test verifies this validation:

  EXPECT(test::throws([&] { h.eval_dim({{lit(5), 10}}); }));  // missing test
  EXPECT(test::throws([&] { h.subs({{h + 1, lit(5)}}); }));   // missing test


  ────────────────────────────────────────
```
* Several of these cases look like they could break expression equality testing. Not exactly an issue with the way we're going to be using these at first however.
* Floor division is a misnomer. Should be truncating division.

**Comment by shivadbhavsar** [comment]
> I asked Claude to examine the code for edge cases:
> 
> ...

Most of these are reasonable asks and so I've gone ahead and implemented/added extra tests

The issue identified in the first point is valid but also intentionally left that way:
- mainly because shape propagation doesn't need this sort of comparison
- expansion would blow up something like (x+2)(y+3)(z-2) to 8 terms
- Paul's more extendible implementation will probably fix this scenario

File: src/include/migraphx/sym.hpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+MIGRAPHX_EXPORT expr parse(const std::string& s);
+
+inline expr operator+(const expr& a, int64_t b) { return a + lit(b); }
+inline expr operator+(int64_t a, const expr& b) { return lit(a) + b; }
+inline expr operator-(const expr& a, int64_t b) { return a - lit(b); }
+inline expr operator-(int64_t a, const expr& b) { return lit(a) - b; }
+inline expr operator*(const expr& a, int64_t b) { return a * lit(b); }
+inline expr operator*(int64_t a, const expr& b) { return lit(a) * b; }
+inline expr operator/(const expr& a, int64_t b) { return a / lit(b); }
+inline expr operator/(int64_t a, const expr& b) { return lit(a) / b; }
```
Comment (pfultz2) [inline src/include/migraphx/sym.hpp]: These should be `friend` functions.

File: src/sym.cpp [hunk 2]
```diff
@@ -0,0 +1,1111 @@
...
+
+using expr_data = std::variant<integer_data, symbol_data, add_data, mul_data, tdiv_data>;
+
+template <class... Ts>
+struct overloaded : Ts...
+{
+    using Ts::operator()...;
+};
+template <class... Ts>
+overloaded(Ts...) -> overloaded<Ts...>;
```
Comment (pfultz2) [inline src/sym.cpp]: This should be moved to the functional.hpp header.

File: src/sym.cpp [hunk 3]
```diff
@@ -0,0 +1,1111 @@
...
+    {
+        h = hash_combine(h, key->cached_hash);
+        h = hash_combine(h, std::hash<int64_t>{}(val));
+    }
+    return h;
+}
+
+static std::size_t compute_hash(const expr_data& d)
+{
+    std::size_t h = std::hash<int>{}(static_cast<int>(d.index()));
```
Comment (pfultz2) [inline src/sym.cpp]: I dont think the `static_cast` is needed here.

File: src/sym.cpp [hunk 4]
```diff
@@ -0,0 +1,1111 @@
...
+    for(const auto& [k, v] : symbol_map)
+    {
+        if(k.empty() or not holds<symbol_data>(k.p->node))
+            MIGRAPHX_THROW("sym::expr::eval_uint: map key '" + k.to_string() + "' is not a symbol");
+        bindings[k.p->node] = static_cast<int64_t>(v);
+    }
+    auto v = eval_direct(p->node, bindings);
+    if(v < 0)
+        MIGRAPHX_THROW("sym::expr::eval_uint: expression evaluated to negative value");
+    return static_cast<std::size_t>(v);
```
Comment (pfultz2) [inline src/sym.cpp]: Remove the `static_cast`.

File: src/sym.cpp [hunk 5]
```diff
@@ -0,0 +1,1111 @@
...
+std::size_t expr::eval_uint(const std::unordered_map<expr, std::size_t>& symbol_map) const
+{
+    if(empty())
+        return 0;
+    binding_map bindings;
+    for(const auto& [k, v] : symbol_map)
+    {
+        if(k.empty() or not holds<symbol_data>(k.p->node))
+            MIGRAPHX_THROW("sym::expr::eval_uint: map key '" + k.to_string() + "' is not a symbol");
+        bindings[k.p->node] = static_cast<int64_t>(v);
```
Comment (pfultz2) [inline src/sym.cpp]: Remove the `static_cast`.

File: src/sym.cpp [hunk 6]
```diff
@@ -0,0 +1,1111 @@
...
+        return make_neg(parse_unary(p));
+    }
+    return parse_primary(p);
+}
+
+static expr_ptr parse_power(const char*& p)
+{
+    auto base = parse_unary(p);
+    skip_ws(p);
+    if(*p == '*' and *(p + 1) == '*')
```
Comment (pfultz2) [inline src/sym.cpp]: pow should not be represented as 2 stars. It should `pow(x, y)`.

File: src/sym.cpp [hunk 7]
```diff
@@ -0,0 +1,1111 @@
...
+    skip_ws(p);
+    if(*p == '-')
+    {
+        ++p;
+        return make_neg(parse_unary(p));
+    }
+    return parse_primary(p);
+}
+
+static expr_ptr parse_power(const char*& p)
```
Comment (pfultz2) [inline src/sym.cpp]: This should be removed. There is no support for pow functions.

File: src/sym.cpp [hunk 8]
```diff
@@ -0,0 +1,1111 @@
...
+    }
+    if(std::isalpha(static_cast<unsigned char>(*p)) != 0 or *p == '_')
+    {
+        std::string name;
+        while(std::isalnum(static_cast<unsigned char>(*p)) != 0 or *p == '_')
+        {
+            name += *p;
+            ++p;
+        }
+        if(name == "floor")
```
Comment (pfultz2) [inline src/sym.cpp]: This should be removed. There is no support for `floor` function.

File: src/sym.cpp [hunk 9]
```diff
@@ -0,0 +1,1111 @@
...
+        if(k.empty() or not holds<symbol_data>(k.p->node))
+            MIGRAPHX_THROW("sym::expr::subs: map key '" + k.to_string() + "' is not a symbol");
+        bindings[k.p->node] = v.p ? v.p->node : make_integer(0);
+    }
+    return {std::make_shared<impl>(substitute(p->node, bindings))};
+}
+
+expr operator+(const expr& a, const expr& b)
+{
+    if(a.empty() and b.empty())
```
Comment (pfultz2) [inline src/sym.cpp]: This should use `or`.

File: src/sym.cpp [hunk 10]
```diff
@@ -0,0 +1,1111 @@
...
+        return {};
+    auto ea = a.p ? a.p->node : make_integer(0);
```
Comment (pfultz2) [inline src/sym.cpp]: Empty expressions shouldnt be treated as 0.

**Review by pfultz2 — APPROVED** [review]

**Review by CharlieL7 — APPROVED** [review]
